Src/Newtonsoft.Json/JsonTextReader.Async.cs

```diff
@@ -0,0 +1,2327 @@
+﻿#region License
+// Copyright (c) 2007 James Newton-King
+//
+// Permission is hereby granted, free of charge, to any person
+// obtaining a copy of this software and associated documentation
+// files (the "Software"), to deal in the Software without
+// restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+#endregion
+
+#if !(NET20 || NET35 || NET40 || PORTABLE40)
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading;
+#if !(NET20 || NET35 || PORTABLE40 || PORTABLE) || NETSTANDARD1_1
+using System.Numerics;
+#endif
+using System.Threading.Tasks;
+using Newtonsoft.Json.Serialization;
+using Newtonsoft.Json.Utilities;
+
+namespace Newtonsoft.Json
+{
+    public partial class JsonTextReader
+    {
+        private bool SafeAsync => GetType() == typeof(JsonTextReader);
```

Comment:
The purpose is that e.g. say someone overrides `ReadAsBoolean`. Now `ReadAsBooleanAsync` which is meant to be an asych equivalent to `ReadAsBoolean` is not, because what `ReadAsBoolean` does has changed. Therefore our implementation of `ReadAsBooleanAsync` asynchronously does the wrong thing. We could use reflection to check on whether either `ReadAsBoolean` and any virtual method it calls into was overridden but just assuming the async methods here aren't viable in a derived class is simpler and faster.
Unsure about the pros and cons of 1) checking once on construction 2) caching on first access and 3) not caching. 2 seems a reasonable middle-ground so I will change to that.